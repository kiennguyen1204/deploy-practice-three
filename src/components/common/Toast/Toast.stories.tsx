import { StoryObj, Meta } from '@storybook/react';

// components
import Toast from 'components/common/Toast';

export default {
  title: 'Components/Toast',
  component: Toast
} as Meta;

type Story = StoryObj<typeof Toast>;

export const Success: Story = {
  args: {
    status: 'success',
    message: 'Success message',
    onClose: () => {}
  }
};

export const Error: Story = {
  args: {
    status: 'error',
    message: 'Error message',
    onClose: () => {}
  }
};
