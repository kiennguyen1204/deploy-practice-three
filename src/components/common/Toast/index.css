.toast {
  position: fixed;
  top: 20px;
  right: 20px;
}

.notification {
  background-color: var(--color-neutral-white);
  border-radius: var(--border-radius-small);
  box-shadow: var(--shadow);
  padding: 10px;
}

.notification .status {
  font-weight: bold;
  margin-bottom: 5px;
}

.notification .content {
  margin: 0;
}

.btn-close {
  background: none;
  border: none;
  color: var(--color-neutral-dark);
  cursor: pointer;
  font-size: var(--font-size-normal);
  margin-left: 10px;
  padding: 0;
}

.btn-close:hover {
  color: var(--color-other-red);
}

.toast:not(.showing):not(.show) {
  opacity: 1 !important;
}

.toast.success .notification {
  background-color: var(--color-success);
}

.toast.success .check-icon {
  color: var(--color-check);
}

.toast.error .notification {
  background-color: var(--color-error);
}

.toast.error .error-icon {
  color: var(--color-error);
}
