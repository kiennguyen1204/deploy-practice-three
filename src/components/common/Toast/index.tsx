import { ReactElement, memo, useEffect } from 'react';

// constants
import { STATUS } from 'constants/enums';

// styles
import './index.css';

interface Props {
  status: string;
  message: string;
  onClose: () => void;
  timeout?: number;
}

const Toast = ({ status, message, onClose, timeout = 2000 }: Props): ReactElement => {
  const toastStatus = status === STATUS.SUCCESS ? 'success' : 'error';
  const toastProcess = status === STATUS.SUCCESS ? 'done' : 'warning';

  useEffect(() => {
    const timer = setTimeout(() => {
      onClose();
    }, timeout);

    return () => {
      clearTimeout(timer);
    };
  }, [timeout, onClose]);

  return (
    <div className={`toast ${toastStatus}`}>
      <div className="notification">
        <span className={`material-symbols-outlined ${toastProcess}`}></span>
        <div className="message">
          <p className="content">{message}</p>
        </div>
        <button onClick={onClose} type="button" className="btn-close" aria-label="Close">
          X
        </button>
      </div>
    </div>
  );
};

export default memo(Toast);
