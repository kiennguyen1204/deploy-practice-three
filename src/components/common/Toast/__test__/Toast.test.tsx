import { fireEvent, render, screen } from '@testing-library/react';

// components
import Toast from 'components/common/Toast';

const defaultProps = {
  status: 'success',
  message: 'Success message',
  onClose: jest.fn()
};

jest.useFakeTimers();

const setup = (props = {}) => {
  const mergedProps = { ...defaultProps, ...props };

  return render(<Toast {...mergedProps} />);
};

describe('Toast component', () => {
  const onCloseMock = jest.fn();
  const clearTimeoutSpy = jest.spyOn(global, 'clearTimeout');

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders success toast correctly', () => {
    const { container, getByText } = setup();

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveClass('success');
    expect(getByText('Success message')).toBeInTheDocument();
  });

  it('renders error toast correctly', () => {
    const { container, getByText } = setup({ status: 'error', message: 'Error message' });

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveClass('error');
    expect(getByText('Error message')).toBeInTheDocument();
  });

  it('calls onClose when close button is clicked', () => {
    const { getByLabelText } = setup({ onClose: onCloseMock });
    const closeButton = getByLabelText('Close');

    fireEvent.click(closeButton);

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('should render the toast message with proper content', () => {
    const status = 'success';
    const message = 'This is a success message';

    render(<Toast status={status} message={message} onClose={onCloseMock} />);
    expect(screen.getByText(message)).toBeInTheDocument();
  });

  it('should call the onClose function after the specified timeout', () => {
    const timeout = 2000;

    render(<Toast status="success" message="Success!" onClose={onCloseMock} timeout={timeout} />);
    jest.advanceTimersByTime(timeout);

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('should clear the timer when unmounting', () => {
    const timeout = 2000;
    const { unmount } = render(
      <Toast status="success" message="Success" onClose={onCloseMock} timeout={timeout} />
    );

    unmount();

    expect(clearTimeoutSpy).toHaveBeenCalledWith(expect.any(Number));
  });

  it('should call the onClose function when close button is clicked', () => {
    const timeout = 2000;

    render(<Toast status="success" message="Success!" onClose={onCloseMock} timeout={timeout} />);
    fireEvent.click(screen.getByLabelText('Close'));

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });
});
