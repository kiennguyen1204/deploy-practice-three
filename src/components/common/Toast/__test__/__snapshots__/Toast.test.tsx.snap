// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component renders error toast correctly 1`] = `
<div
  class="toast error"
>
  <div
    class="notification"
  >
    <span
      class="material-symbols-outlined warning"
    />
    <div
      class="message"
    >
      <p
        class="content"
      >
        Error message
      </p>
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    >
      X
    </button>
  </div>
</div>
`;

exports[`Toast component renders success toast correctly 1`] = `
<div
  class="toast success"
>
  <div
    class="notification"
  >
    <span
      class="material-symbols-outlined done"
    />
    <div
      class="message"
    >
      <p
        class="content"
      >
        Success message
      </p>
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    >
      X
    </button>
  </div>
</div>
`;
